presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/SpatialPartitioningInternalAggregateFunction.java

```diff
@@ -0,0 +1,108 @@
+
+import static com.facebook.presto.geospatial.KDBTree.buildKDBTree;
+import static com.facebook.presto.geospatial.serde.GeometrySerde.deserializeEnvelope;
+import static com.facebook.presto.plugin.geospatial.GeometryType.GEOMETRY_TYPE_NAME;
+import static com.facebook.presto.spi.type.StandardTypes.INTEGER;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static java.lang.Math.floor;
+import static java.lang.Math.toIntExact;
+
+@AggregationFunction(value = "spatial_partitioning", decomposable = false, isOrderSensitive = true, hidden = true)
+public class SpatialPartitioningInternalAggregateFunction
+{
+    private static final JsonCodec<KDBTree> KDB_TREE_CODEC = new JsonCodecFactory().jsonCodec(KDBTree.class);
+    private static final int MAX_SAMPLE_COUNT = 1_000_000;
+
+    private SpatialPartitioningInternalAggregateFunction() {}
+
+    @InputFunction
+    public static void input(SpatialPartitioningState state, @SqlType(GEOMETRY_TYPE_NAME) Slice slice, @SqlType(INTEGER) long partitionCount)
+    {
+        Envelope envelope = deserializeEnvelope(slice);
+        if (envelope == null) {
+            return;
+        }
+
+        Rectangle extent = new Rectangle(envelope.getXMin(), envelope.getYMin(), envelope.getXMax(), envelope.getYMax());
+
+        if (state.getCount() == 0) {
+            state.setPartitionCount(toIntExact(partitionCount));
+            state.setExtent(extent);
+            state.setSamples(new ArrayList<>());
+        }
+        else {
+            state.setExtent(state.getExtent().merge(extent));
+        }
+
+        List<Rectangle> samples = state.getSamples();
+        if (samples.size() <= MAX_SAMPLE_COUNT) {
+            samples.add(extent);
+        }
+        else {
+            int sampleIndex = (int) floor(ThreadLocalRandom.current().nextDouble() * state.getCount());
+            if (sampleIndex < MAX_SAMPLE_COUNT) {
+                samples.set(sampleIndex, extent);
+            }
+        }
+
+        state.setCount(state.getCount() + 1);
+    }
+
+    @CombineFunction
+    public static void combine(SpatialPartitioningState state, SpatialPartitioningState otherState)
+    {
+        throw new UnsupportedOperationException("spatial_partitioning must run on a single node");
+    }
+
+    @OutputFunction(StandardTypes.VARCHAR)
+    public static void output(SpatialPartitioningState state, BlockBuilder out)
+    {
+        if (state.getCount() == 0) {
+            out.appendNull();
+            return;
+        }
+
+        List<Rectangle> samples = state.getSamples();
+
+        int partitionCount = state.getPartitionCount();
+        int maxItemsPerNode = (int) (Math.ceil(samples.size() * 1.0 / partitionCount));
+        Rectangle envelope = state.getExtent();
+
+        // Add a small buffer on the right and upper sides
+        Rectangle paddedExtent = new Rectangle(envelope.getXMin(), envelope.getYMin(), envelope.getXMax() + 0.01, envelope.getYMax() + 0.01);
```

Comment:
What's the reason of adding this padding? Is it double overflow?

Reply:
@arhimondr This is to allow using half-open rectangles to ensure that any point falls into exactly one partition.

Reply:
@martint In literature, KDB tree is defined in terms of proper rectangles. I find it very hard to think though what it means for a KDB tree made of half-open rectangles. At the same time, this way of building a KDB tree is straightforward and allows to avoid duplicating points unnecessarily and skip de-dup logic in the join.

Reply:
Fair enough. The problem I have with this is the seemingly arbitrary choice of 0.01 as the padding and what it implies (what if all geometries being processed are smaller than 0.01 units?) If all we need is a number that's higher than XMax, we should probably use `Math.nextUp` to calculate it.